knowledge_storm/collaborative_storm/engine.py

```diff
@@ -0,0 +1,691 @@
+        )
+        return costorm_runner
+
+    def warm_start(self):
+        with self.logging_wrapper.log_pipeline_stage(
+            pipeline_stage=f"warm start stage"
+        ):
+            if not self.runner_argument.rag_only_baseline_mode:
+                warm_start_module = WarmStartModule(
+                    lm_config=self.lm_config,
+                    runner_argument=self.runner_argument,
+                    logging_wrapper=self.logging_wrapper,
+                    rm=self.rm,
+                    callback_handler=self.callback_handler,
+                )
+
+                warmstart_conv, warmstart_revised_conv, warmstart_experts = (
+                    warm_start_module.initiate_warm_start(
+                        topic=self.runner_argument.topic,
+                        knowledge_base=self.knowledge_base,
+                    )
+                )
+                self.discourse_manager.experts = (
+                    self.discourse_manager._parse_expert_names_to_agent(
+                        warmstart_experts
+                    )
+                )
+                self.discourse_manager.next_turn_moderator_override = True
+                self.conversation_history = (
+                    warmstart_revised_conv if warmstart_revised_conv else warmstart_conv
+                )
+                self.warmstart_conv_archive = warmstart_conv
+                self.knowledge_base.reogranize()
+            else:
+                if self.knowledge_base is None:
+                    self.knowledge_base = KnowledgeBase(
+                        topic=self.runner_argument.topic
+                    )
+                if self.conversation_history is None:
+                    self.conversation_history = []
+                conv_turn = (
+                    self.discourse_manager.pure_rag_agent.generate_topic_background()
+                )
+                self.conversation_history.append(conv_turn)
+                self.knowledge_base.update_from_conv_turn(
+                    conv_turn=conv_turn,
+                    allow_create_new_node=True,
+                    insert_under_root=self.runner_argument.rag_only_baseline_mode,
+                )
+
+    def generate_report(self) -> str:
+        with self.logging_wrapper.log_pipeline_stage("report generation stage"):
+            with self.logging_wrapper.log_event(
+                "report generation stage: generate report"
+            ):
+                return self.knowledge_base.to_report()
+
+    def dump_logging_and_reset(self):
+        return self.logging_wrapper.dump_logging_and_reset()
+
+    def get_next_turn_experts_for_frontend_render(self):
+        turn_policy = self.discourse_manager.get_next_turn_policy(
+            conversation_history=self.conversation_history,
+            simulate_user=False,
+            dry_run=True,
+        )
+        return turn_policy.agent.role_name
+
+    def step(
```

Comment:
Add a comment for this method.

[Here](https://github.com/openai/gym/blob/master/gym/core.py#L87) is a good example for informative comment.